crowbar_framework/app/models/service_object.rb

```diff
@@ -484,12 +545,22 @@ def process_queue
       # For each ready item, apply it.
       list.each do |item|
         @logger.debug("process queue: item to do: #{item.inspect}")
+
         bc = item["barclamp"]
         inst = item["inst"]
         service = eval("#{bc.camelize}Service.new @logger")
+
+        # This will call apply_role and chef-client.
+        # Params: (inst, in_queue, validate_after_save)
         answer = service.proposal_commit(inst, true, false)
+
         @logger.debug("process queue: item #{item.inspect}: results #{answer.inspect}")
+
+        # 202 means some nodes are not ready, bail out in that case
+        # FIXME: we're re-running the whole apply continuously, until there
+        # are no items left in the queue?
```

Comment:
Yes. And it's lame because from the user perspective, we're still applying the first barclamp while this part was completed and we're actually applying what was in the queue.

This needs to be fixed.